### PR TITLE
Allows usage of public ip, private ip or hostname in Linode inventory

### LINKED ISSUE #51195: Support using public ip in the v4 Linode dynamic inventory plugin
##### SUMMARY
The older Linode dynamic inventory plugin allowed the public IP to be used as well as the instance's label. Bring that functionality back to the newer v4 compatible Linode inventory plugin.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
`lib/ansible/plugins/inventory/linode.py`

##### ADDITIONAL INFORMATION
This allows the inventory plugin to be used with infrastruction that does not have DNS hostnames or hostnames that match Linode labels. 

From the older plugin....
```
NOTE: By default, this script also assumes that the Linodes in your account all have
labels that correspond to hostnames that are in your resolver search path.
Your resolver search path resides in /etc/hosts.
Optionally, if you would like to use the hosts public IP instead of it's label use
the following setting in linode.ini:
    use_public_ip = true
```

### PR DESCRIPTION
**Pull request by mcg**
The older linode inventory plugin supported using the public IP as well as the
instances label. This change enables using public IP, private IP or hostname. It also injects public_ip and private_ip into `hostvars` for easy access(also as the old plugin).

##### SUMMARY
Fixes #51195 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
`lib/ansible/plugins/inventory/linode.py`

### CODE REVIEW

**Comment by Akasurde** [comment]
CI failure -  https://app.shippable.com/github/ansible/ansible/runs/103323/6/console
```
00:18 =================================== FAILURES ===================================
00:18 _______________________ test_empty_config_query_options ________________________
00:18 [gw1] linux2 -- Python 2.7.12 /usr/bin/python2.7
00:18 inventory = <ansible.plugins.inventory.linode.InventoryModule object at 0x7f45f85cb950>
00:18 
00:18     def test_empty_config_query_options(inventory):
00:18 >       regions, types = inventory._get_query_options({})
00:18 E       AttributeError: 'InventoryModule' object has no attribute '_get_query_options'
00:18 
00:18 test/units/plugins/inventory/test_linode.py:61: AttributeError
00:18 ___________________________ test_conig_query_options ___________________________
00:18 [gw1] linux2 -- Python 2.7.12 /usr/bin/python2.7
00:18 inventory = <ansible.plugins.inventory.linode.InventoryModule object at 0x7f45f85cb950>
00:18 
00:18     def test_conig_query_options(inventory):
00:18 >       regions, types = inventory._get_query_options({
00:18             'regions': ['eu-west', 'us-east'],
00:18             'types': ['g5-standard-2', 'g6-standard-2'],
00:18         })
00:18 E       AttributeError: 'InventoryModule' object has no attribute '_get_query_options'
00:18 
00:18 test/units/plugins/inventory/test_linode.py:66: AttributeError
00:18 --- generated xml file: /root/ansible/test/results/junit/python2.7-units.xml ---
00:18 =========================== short test summary info ============================
00:18 FAIL test/units/plugins/inventory/test_linode.py::test_empty_config_query_options
00:18 FAIL test/units/plugins/inventory/test_linode.py::test_conig_query_options
```

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]: Can we name `dest` better? I feel like `host_identifier` (or so) might be more readable.

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]: `%s/a the/the`

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]: I think we should guard against an `IndexError`. Even if Linode guarantee us some value, given programming error and such, I'd like to guarantee users of this module error messages that lead them in the right direction.

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]: Again, can we name this `dest` better? Will help readability as the module grows.

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]: Did you experience some bug with the `endings` detection or?

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py:199]: I'm not sure why you changed this? It could be described both ways, I suppose. 

Hence, this feels like it could be bikeshed :sweat_smile:.

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]: Thinking ahead, in case there are other ways to specify the identifier for an instance (lable, ip, etc.), we might implement here a `self._determine_host_identifier` or so, which negotiates the options and decides if we're using the `public_ip` or the `label` and then passes that to the `_add_instances_to_groups` and `_add_hostvars_for_instances`? We could then eliminate the `if/else` clause in those function sand keep the control flow more simple. WDYT?

**Review by decentral1se — CHANGES_REQUESTED** [review]
Thanks for this @mcg! I've asked other Linode working group members to review and keep an eye on this, so we should be getting it in promptly. We're lurking on freenode/#ansible-linode if you have any questions. Otherwise, just ping here.

File: lib/ansible/plugins/inventory/linode.py
Comment (mcg) [reply lib/ansible/plugins/inventory/linode.py:199]: "query" felt like this was only related to configuring the API query, which it was with region/type. Adding `use_public_ip` isn't related to query, just config.

File: lib/ansible/plugins/inventory/linode.py
Comment (mcg) [reply lib/ansible/plugins/inventory/linode.py]: Will change

File: lib/ansible/plugins/inventory/linode.py
Comment (mcg) [reply lib/ansible/plugins/inventory/linode.py]: Oops. Will fix.

File: lib/ansible/plugins/inventory/linode.py
Comment (mcg) [reply lib/ansible/plugins/inventory/linode.py]: No, just my local env out of sync. Fixed.

File: lib/ansible/plugins/inventory/linode.py
Comment (mcg) [reply lib/ansible/plugins/inventory/linode.py]: Added some guarding here. Hopefully what you're going for.

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [reply lib/ansible/plugins/inventory/linode.py:199]: Got it, agreed!

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]: This should use `AnsibleError` like the other errors to pass back a nicely formatted message to the user.

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]: I don't understand why we would `return False` here. This function is about giving back an instance based on the public IP address or failing gracefully with an error message for the user. If this `return False` is happening, we need to write code to handle that in all places it is called within the module. This can lead to extra `if/else` which we should avoid for maintainability purposes.

File: test/units/plugins/inventory/test_linode.py
Comment (decentral1se) [inline test/units/plugins/inventory/test_linode.py]: Oh, we should `%s/conig/config/`

**Review by decentral1se — CHANGES_REQUESTED** [review]
Thanks for pushing on with this @mcg. I've given it another pass.

File: lib/ansible/plugins/inventory/linode.py
Comment (mcg) [reply lib/ansible/plugins/inventory/linode.py]: Wasn't sure about using that here, will change.

File: test/units/plugins/inventory/test_linode.py
Comment (mcg) [reply test/units/plugins/inventory/test_linode.py]: sigh.. will fix

File: lib/ansible/plugins/inventory/linode.py
Comment (mcg) [reply lib/ansible/plugins/inventory/linode.py]: Hmm ok. Will rework this.

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]: Could we get a few tests for the different flows in this function?

**Review by decentral1se — CHANGES_REQUESTED** [review]
Final one! :fire: :fire: :fire:

**Comment by decentral1se** [comment]
Bump! Any news on this @mcg? Thanks for the efforts :100:

**Comment by mcg** [comment]
@decentral1se Trying to get back to this, what we're using internally has diverged a bit from this work. Basically, instead of `use_public_ip: true/false`, we're using `instance_access: public_ip/private_ip/hostname` along with injecting `public_ip` and `private_ip` into the inventory as the old plugin also did. Should I push this work to this PR?

**Comment by decentral1se** [comment]
Hey @mcg, I'm not sure I follow, does that map what the Linode API tooling exposes with v4? If so, then sure, I'd be happy to review and push to get merged. Not sure if we want adhoc keys otherwise - but please tell if I have this wrong.

**Comment by mcg** [comment]
@decentral1se the old plugin added keys for public_ip and private_ip(and a few other values), so that you can access and use these in hostvars directly. They're not adhoc so much as just not directly accessible in the API. This is something we use heavily for building firewall rules, loadbalancer backends etc...

Here is what we're doing:
```
    def _add_hostvars_for_instances(self, instance_access):
        """Add hostvars for instances in the dynamic inventory."""
        for instance in self.instances:
            if instance_access == 'public_ip':
                host_identifier = self._get_instance_public_ip(instance)
            elif instance_access == 'private_ip':
                host_identifier = self._get_instance_private_ip(instance)
            elif instance_access == 'hostname':
                host_identifier = instance.label
            else:
                raise AnsibleError(
                    'Instance Access setting is not valid'
                )

            self.inventory.set_variable(
                host_identifier,
                "public_ip",
                self._get_instance_public_ip(instance)
            )
            self.inventory.set_variable(
                host_identifier,
                "private_ip",
                self._get_instance_private_ip(instance)
            )
            hostvars = instance._raw_json
            for hostvar_key in hostvars:
                self.inventory.set_variable(
                    host_identifier,
                    hostvar_key,
                    hostvars[hostvar_key]
                )
```

**Comment by decentral1se** [comment]
Oh yes, that looks powerful and useful, let's have it! 

You can rename this PR + wipe the changes or open another, as you want. 

Just ping and I'll happily review. :rocket:

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]:
```suggestion
                'public_ip',
```

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]:
```suggestion
                'private_ip',
```

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]: We could merge the two `_get_instance*_ip` functions with something like?

```
def _get_instance_by_ip(self, instance, is_private=False):
    """Returns the IPv4 address of the instance."""
    if instance.ipv4 and len(instance.ipv4) > 0:
        for ip in instance.ipv4:
            parsed_ip = ipaddress.ip_address(ip)

            if parsed_ip.is_private and is_private:
                return ip

            if parsed_ip.is_private and not is_private
                raise AnsibleError('Expected public IP not found')

            return parsed_ip

    raise AnsibleError('Instance IPv4 address is missing or empty')
```

Less code :rocket:

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]: You can use the following docs key to specify what choices are valid:

```
choices:
- hostname
- public_ip
- private_ip
```

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]: If we use `choices` in the documentation fragment, we can avoid this (invalid values will never be allowed here).

**Review by decentral1se — CHANGES_REQUESTED** [review]
Thanks for picking this up again :+1:

File: lib/ansible/plugins/inventory/linode.py
Comment (mcg) [reply lib/ansible/plugins/inventory/linode.py]: How about....
```
  def _get_instance_ip(self, instance, private=False):
        """Returns the public or private IP address of the instance"""
        if instance.ipv4 and len(instance.ipv4) > 0:
            result_ip = None
            for ip in instance.ipv4:
                if not ipaddress.ip_address(ip).is_private and not private:
                    result_ip = ip
                elif ipaddress.ip_address(ip).is_private and private:
                    result_ip = ip
            if result_ip:
                return result_ip
        else:
            raise AnsibleError(
                'Instance IPv4 address is missing or empty'
            )
```

Comment (decentral1se) [reply lib/ansible/plugins/inventory/linode.py]: Yes, great!

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py:144]: Oh woops, are we missing that `raise AnsibleError...` when the `result_ip` is `None`?

File: lib/ansible/plugins/inventory/linode.py
Comment (decentral1se) [inline lib/ansible/plugins/inventory/linode.py]: Same as https://github.com/ansible/ansible/pull/51196#discussion_r266537178, we can drop it?

**Comment by decentral1se** [comment]
Final comments, promise :) Looks good!

File: lib/ansible/plugins/inventory/linode.py
Comment (mcg) [reply lib/ansible/plugins/inventory/linode.py:144]: I keep forgetting private ip is optional. Added that back and a test.

File: lib/ansible/plugins/inventory/linode.py
Comment (mcg) [reply lib/ansible/plugins/inventory/linode.py]: Drop what? Even using "choice" we have to set the `host_identifier` based on the choice?

Comment (decentral1se) [reply lib/ansible/plugins/inventory/linode.py]: Ah, right, got it :+1:

**Review by decentral1se — APPROVED** [review]
:rocket: 

shipit

**Comment by mcg** [comment]
@decentral1se Thanks for all the help!

**Comment by decentral1se** [comment]
@InTheCloudDan @displague @rmcintosh, can we get some shipit gold on this one?

**Comment by InTheCloudDan** [comment]
shipit

**Comment by decentral1se** [comment]
Looks like `core_review` shouldn't be marked for this issue (see https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#core). I've gone and nagged on `#ansible-devel` to get it pushed through.

**Comment by gundalow** [comment]
bot_status
Wondering where `support:core` is being set

**Comment by rmcintosh** [comment]
shipit

File: lib/ansible/plugins/inventory/linode.py
Comment (bcoca) [inline lib/ansible/plugins/inventory/linode.py]: Instead of hardcoding this choice, why not use constructed/compose functionality to allow user to set `ansible_host` and similar variables to anything they need.

Comment (mcg) [reply lib/ansible/plugins/inventory/linode.py]: Can you explain a bit more how this would work? I've dug around a bit and am not sure.

Comment (bcoca) [reply lib/ansible/plugins/inventory/linode.py]: basically #53869 does same for vultr plugin

**Comment by decentral1se** [comment]
FYI @mcg, you've got until tomorrow to get this merged or it will be frozen out due to https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_8.rst. No pressure :)

File: lib/ansible/plugins/inventory/linode.py
Comment (mcg) [reply lib/ansible/plugins/inventory/linode.py]: Does make sense to use `ansible_host` rather than change the host identifier. I'll make that change.

**Comment by mcg** [comment]
When is the next window to get this merged, after 2019-05-16 ?

File: lib/ansible/plugins/inventory/linode.py
Comment (bcoca) [reply lib/ansible/plugins/inventory/linode.py]: yes, since that is what you specifically are doing with the new option 'setting ansible_host' while the compose system gives you a generic way of setting any/all kinds of variables, you really only need this type of option for 'inventory_hostname'

Comment (mcg) [reply lib/ansible/plugins/inventory/linode.py]: Thanks for following up, this does use `ansible_host` now.

**Comment by decentral1se** [comment]
I'm not sure but you can quickly find out on the #ansible-dev channel or it is probably documented somewhere on the ROADMAP* files in the documentation. As the `core_review` label is on this now, I won't be able to get this merged from the Linode team side. So, your best bet is to push for core devs (ones commenting here and some on #ansible-dev channel) to get it in. Hope that helps! Thanks for the work.

**Comment by mcg** [comment]
@decentral1se Thanks again for the efforts and help.

File: lib/ansible/plugins/inventory/linode.py
Comment (bcoca) [reply lib/ansible/plugins/inventory/linode.py]: maybe I am not being clear, my point is you do NOT need THIS specific option, using compose any user can decide exactly what `ansible_host` value is set to, as well as any other connection or non connection specific variable.

**Review by bcoca — CHANGES_REQUESTED** [review]
use compose instead of having to add specific options per connection variable

File: lib/ansible/plugins/inventory/linode.py
Comment (s-hertel) [inline lib/ansible/plugins/inventory/linode.py:44]: These are already included via the `constructed` fragment. You can remove these two lines and check by using `ansible-doc -t inventory linode` on your branch.
```suggestion

```

File: lib/ansible/plugins/inventory/linode.py
Comment (s-hertel) [inline lib/ansible/plugins/inventory/linode.py:45]:
```suggestion

```

File: lib/ansible/plugins/inventory/linode.py
Comment (s-hertel) [inline lib/ansible/plugins/inventory/linode.py:64]: This should probably have a documented requirement or you should fall back to using `ansible.module_utils.compat.ipaddress` if there's an ImportError

**Comment by mkrizek** [comment]
Thank you very much for your interest in Ansible. This plugin is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general.

Please re-submit this issue in the above repository.

If you have further questions please stop by IRC or the mailing list:

* IRC: #ansible on irc.freenode.net
* mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Pull request closed without merging**
